miri-script/src/commands.rs

```diff
@@ -0,0 +1,711 @@
+    ) -> Result<()> {
+        cmd(
+            "cargo",
+            chain!(
+                [&toolchain, "build", extra_flags, "--manifest-path", path,],
+                args.into_iter().cloned()
+            ),
+        )
+        .run()?;
+        Ok(())
+    }
+
+    fn build(&self, flags: &[String]) -> Result<()> {
+        // Build, and let caller control flags.
+        let toolchain = self.get_toolchain_override();
+        let miri_manifest = path!(self.miri_dir / "Cargo.toml");
+        let cargo_miri_manifest = path!(self.miri_dir / "cargo-miri" / "Cargo.toml");
+        let args = flags.iter().map(String::as_str).collect::<Vec<_>>();
+        Self::build_package(
+            &miri_manifest.to_string_lossy().to_owned(),
+            &toolchain,
+            &self.cargo_extra_flags,
+            &args,
+        )?;
+        Self::build_package(
+            &cargo_miri_manifest.to_string_lossy().to_owned(),
+            &toolchain,
+            &self.cargo_extra_flags,
+            &args,
+        )?;
+        Ok(())
+    }
+
+    fn check(&self, flags: &[String]) -> Result<()> {
+        fn check_package(
+            // Path to Cargo.toml file of a package to check.
+            path: &str,
+            toolchain: &str,
+            extra_flags: &str,
+            all_targets: bool,
+            args: &[&str],
+        ) -> Result<()> {
+            let all_targets = all_targets.then_some("--all-targets");
+            cmd(
+                "cargo",
+                chain!(
+                    [&toolchain, "check", extra_flags, "--manifest-path", path,],
+                    all_targets,
+                    args.into_iter().cloned()
+                ),
+            )
+            .run()?;
+            Ok(())
+        }
+        // Check, and let caller control flags.
+        let toolchain = self.get_toolchain_override();
+        let miri_manifest = path!(self.miri_dir / "Cargo.toml");
+        let cargo_miri_manifest = path!(self.miri_dir / "cargo-miri" / "Cargo.toml");
+        let args = flags.iter().map(String::as_str).collect::<Vec<_>>();
+        check_package(
+            &miri_manifest.to_string_lossy().to_owned(),
+            &toolchain,
+            &self.cargo_extra_flags,
+            true,
+            &args,
+        )?;
+        check_package(
+            &cargo_miri_manifest.to_string_lossy().to_owned(),
+            &toolchain,
+            &self.cargo_extra_flags,
+            false,
+            &args,
+        )?;
+        Ok(())
+    }
+
+    /// Returns a toolchain override for use in command line (`+{toolchain_name}`).
+    fn get_toolchain_override(&self) -> String {
+        format!("+{}", &self.active_toolchain)
+    }
+    fn test(&mut self, flags: &[String]) -> Result<()> {
+        let bless = matches!(self.command, Subcommands::Bless { .. });
+        let miri_manifest = path!(self.miri_dir / "Cargo.toml");
+        let miri_manifest = miri_manifest.to_string_lossy().to_owned();
```

Comment:
As @saethlin noted, ultimately we are doing to pass this to a command-running thing so we shouldn't need this.

The standard `Command` works with all `AsRef<OsStr>`, so a `PathBuf` like this could just be used without any further conversion. Is it possible to achieve the same here?

Reply:
Noted, I've kind of addressed that in follow-up commits. The issue I am facing is that when I'm passing args to `cmd`, some of them are string literals, whereas others are `OsStr`s already. This forces me to convert some arguments and make all of this unnecessarily noisy. I wish there was a way to e.g. accept a tuple of types that all implement `AsRef<OsStr` in chain :/

Reply:
Yeah I was hoping `duct` would provide that...

Can't we have our own `chain_cmd` function with a signature like this?
```rust
fn chain_cmd(a: impl IntoIterator<Item = impl AsRef<OsStr>>, b: impl IntoIterator<Item = impl AsRef<OsStr>>) -> impl Iterator<Item=&OsStr>
```

Reply:
I guess this is not valid anymore after a move to `xshell`?